plugins/home/src/components/CustomHomepageGrid.tsx

```diff
@@ -0,0 +1,205 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React from 'react';
+import { Layout, Layouts, Responsive, WidthProvider } from 'react-grid-layout';
+import {
+  componentDataKey,
+  MaybeComponentNode,
+  storageApiRef,
+  useApi,
+} from '@backstage/core-plugin-api';
+import 'react-grid-layout/css/styles.css';
+// eslint-disable-next-line import/no-extraneous-dependencies
+import 'react-resizable/css/styles.css';
+import { SpeedDial } from '@material-ui/lab';
+import SpeedDialIcon from '@material-ui/lab/SpeedDialIcon';
+import {
+  createStyles,
+  ListItemSecondaryAction,
+  makeStyles,
+  Modal,
+  Theme,
+} from '@material-ui/core';
+import ListItemText from '@material-ui/core/ListItemText';
+import List from '@material-ui/core/List';
+import ListItem from '@material-ui/core/ListItem';
+import { compact } from 'lodash';
+import IconButton from '@material-ui/core/IconButton';
+import AddIcon from '@material-ui/icons/Add';
+import { useObservable } from 'react-use';
+
+// eslint-disable-next-line new-cap
+const Grid = WidthProvider(Responsive);
+
+const useStyles = makeStyles((theme: Theme) =>
+  createStyles({
+    speedDial: {
+      position: 'absolute',
+      bottom: theme.spacing(2),
+      right: theme.spacing(2),
+    },
+    modal: {
+      position: 'absolute',
+      width: '50%',
+      backgroundColor: theme.palette.background.paper,
+      border: '2px solid #000',
+      top: '25%',
+      left: '25%',
+      boxShadow: theme.shadows[5],
+      padding: theme.spacing(2, 4, 3),
+    },
+    widgetWrapper: {
+      '& > *:first-child': {
+        width: '100%',
+        height: '100%',
+      },
+    },
+  }),
+);
+
+const getAvailableWidgets = (children: MaybeComponentNode[]) => {
+  const ret = children.map(child => {
+    if (!child.type) {
+      return null;
+    }
+    return {
+      component: child,
+      ...Object.fromEntries(child.type[componentDataKey]?.map as Iterable<any>),
+    } as any;
+  });
+  return compact(ret);
+};
+
+function useHomeStorage(
+  defaultLayout: Layout[],
+): [Layout[], (value: Layout[]) => void] {
+  const storageApi = useApi(storageApiRef);
+  const storageKey = 'home_layout';
+  // TODO: Support multiple layouts
+  const setValue = (value: Layout[]) => {
+    storageApi.set(storageKey, JSON.stringify(value));
+  };
+  const snapshot = useObservable(
+    storageApi.observe$<string>(storageKey),
+    storageApi.snapshot(storageKey),
+  );
+  let result: Layout[];
+  if (snapshot.presence === 'absent') {
+    result = defaultLayout;
+  } else {
+    result = JSON.parse(snapshot.value!);
+  }
+  return [result, setValue];
```

Comment:
in the final code, remember that setValue should be stable and never change (needs a useCallback on it), and the result should only change if the value out of the observable actually changes (needs a useMemo or similar on it)